lightning/src/routing/scoring.rs

```diff
@@ -662,11 +704,21 @@ impl<L: Deref<Target = u64>, T: Time, U: Deref<Target = T>> DirectedChannelLiqui
 		} else {
 			let numerator = max_liquidity_msat + 1 - amount_msat;
 			let denominator = max_liquidity_msat + 1 - min_liquidity_msat;
-			approx::negative_log10_times_1024(numerator, denominator)
-				.saturating_mul(liquidity_penalty_multiplier_msat) / 1024
+			match params.cost_function {
+				ProbabilisticScoringCostFunction::NegativeLogSuccessProbability => {
+					approx::negative_log10_times_1024(numerator, denominator)
+						.saturating_mul(params.liquidity_penalty_multiplier_msat) / 1024
+				},
+				ProbabilisticScoringCostFunction::TwiceFailureProbability => {
+					// Avoid floating-point operations by multiplying the coefficient through:
+					// 2 * liquidity_penalty_multiplier_msat * (1 - success_probability)
+					let coefficient = max_penalty_msat;
+					coefficient.saturating_sub(coefficient.saturating_mul(numerator) / denominator)
+				},
+			}
 		}
 		// Upper bound the penalty to ensure some channel is selected.
-		.min(2 * liquidity_penalty_multiplier_msat)
+		.min(max_penalty_msat)
```

Comment:
Actually, can we drop this? The only case where it should be reachable is the `u64::max_value()` case where we our probability is 0, right?

Reply:
Hmm... I don't think so because it is the difference of two lookups. Would need to increase the lookup table size quite a bit.

Reply:
Ah, right, ugh. Maybe lets just inline this into the `NegativeLogSuccessProbability` case, at least?

Reply:
Ideally, this would stay as any cost function should be limited by that. Assuming you have some performance concern? Moved it to the `NegativeLogSuccessProbability` case.